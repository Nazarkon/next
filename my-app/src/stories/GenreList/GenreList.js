'use client'
import React from 'react';
import PropTypes from 'prop-types';
import './GenreList.scss';

const GenreList = (props) => {
  const { genreList, currentItem, updateList } = props;

  const capitalizeLetter = (word) => {
    return word.charAt(0).toUpperCase() + word.slice(1);
  };

  return (
    <ul className="list-container">
      {genreList.map((item) => (
        <li
          key={item.id}
          className={item.id === currentItem.id ? 'item-light' : ''}
          onClick={() => updateList(item)}
        >
          {capitalizeLetter(item.name)}
        </li>
      ))}
    </ul>
  );
};

GenreList.propTypes = {
  genreList: PropTypes.array.isRequired,
  currentItem: PropTypes.object.isRequired,
  updateList: PropTypes.func
};

export default GenreList;
