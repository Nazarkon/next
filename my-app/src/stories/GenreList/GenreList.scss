.list-container{
    list-style-type: none;
    display: flex;
}
ul{
    padding: 0;
}
li{
    margin: 10px;
    color: white;
}
.bg-salmon {
    background-color: salmon;
    color: white;
  }
  
.item-light {
    transition: color 0.2s ease-in-out;
    border-bottom: 4px solid #F65261;
}