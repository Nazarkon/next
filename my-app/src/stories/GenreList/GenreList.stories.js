import React, { useState } from 'react';

import GenreList from './GenreList';

export default {
  title: 'Components/GenreList',
  component: GenreList,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    genreList: {
      control: 'array',
      description: 'Default genre array',
      defaultValue: [
        {
          id: 'all',
          name: 'All',
          is_active: false
        },
        {
          id: 'documentary',
          name: 'Documentary',
          is_active: false
        },
        {
          id: 'comedy',
          name: 'Comedy',
          is_active: false
        },
        {
          id: 'horror',
          name: 'Horror',
          is_active: false
        },
        {
          id: 'crime',
          name: 'Crime',
          is_active: false
        }
      ]
    },
    currentItem: {
      control: 'object',
      description: 'Example of props object',
      defaultValue: {
        id: 'all',
        name: 'All',
        is_active: false
      }
    }
  }
};

const Template = (args) => {
  const [currentItem, setCurrentItem] = useState(args.currentItem);

  const updateList = (item) => {
    setCurrentItem(item);
  };

  return <GenreList {...args} currentItem={currentItem} updateList={updateList} />;
};

export const Default = Template.bind({});

Default.args = {
  genreList: [
    {
      id: 'all',
      name: 'All',
      is_active: false
    },
    {
      id: 'documentary',
      name: 'Documentary',
      is_active: false
    },
    {
      id: 'comedy',
      name: 'Comedy',
      is_active: false
    },
    {
      id: 'horror',
      name: 'Horror',
      is_active: false
    },
    {
      id: 'crime',
      name: 'Crime',
      is_active: false
    }
  ],
  currentItem: {
    id: 'all',
    name: 'All',
    is_active: false
  }
};
