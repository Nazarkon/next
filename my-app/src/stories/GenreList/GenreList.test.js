import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import GenreList from './GenreList';
import genreList from '../../mock/genreData.json';

describe('Check GenreList element behavior', () => {
  it('Matches snapshot', () => {
    const { container } = render(<GenreList genreList={genreList} currentItem={genreList[0]} />);
    expect(container).toMatchSnapshot();
  });
  test('Check that component renders all genres passed in props', () => {
    const { getByText } = render(<GenreList genreList={genreList} currentItem={genreList[0]} />);

    expect(getByText('All')).toBeInTheDocument();

    expect(getByText('Fantasy')).toBeInTheDocument();

    expect(getByText('Adventure')).toBeInTheDocument();

    expect(getByText('Romance')).toBeInTheDocument();

    expect(getByText('Science Fiction')).toBeInTheDocument();
  });

  test('Check that component highlights a selected genre passed in props', () => {
    const selectedGenre = genreList[0];
    render(<GenreList genreList={genreList} currentItem={selectedGenre} />);

    const capitalizeLetter = (word) => {
      return word.charAt(0).toUpperCase() + word.slice(1);
    };

    const selectedGenreElement = screen.getByText(capitalizeLetter(selectedGenre.name));

    expect(selectedGenreElement).toHaveClass('item-light');
  });

  test('Check that after a click event on a genre button component calls onChange callback and passes correct genre in arguments', () => {
    const selectedGenre = genreList[0];
    const onClick = jest.fn();
    render(<GenreList genreList={genreList} currentItem={selectedGenre} updateList={onClick} />);

    const genreElement = screen.getByText('Romance');

    userEvent.click(genreElement);

    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onClick).toHaveBeenCalledWith({
      id: 'romance',
      is_active: false,
      name: 'romance'
    });
  });
});
