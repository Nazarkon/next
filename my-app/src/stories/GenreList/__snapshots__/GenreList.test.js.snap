// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check GenreList element behavior Matches snapshot 1`] = `
<div>
  <ul
    class="list-container"
  >
    <li
      class="item-light"
    >
      All
    </li>
    <li
      class=""
    >
      Fantasy
    </li>
    <li
      class=""
    >
      Adventure
    </li>
    <li
      class=""
    >
      Romance
    </li>
    <li
      class=""
    >
      Science Fiction
    </li>
  </ul>
</div>
`;
